### ISSUE TITLE
Crash bug in Filesystem

### ISSUE THREAD

**Issue by DanRStevens** [lairworks/nas2d-core#109]
The `Filesystem` object will unconditionally call `PHYSFS_deinit()` in the destructor. However, the corresponding `PHYSFS_init(argv_0)` call is not placed in the constructor, but rather deferred until `init(...)` is called. This means a `Filesystem` object can not be safely destructed until after `init(...)` has been called.

The following may exhibit the crash behavior:
```cpp
{  // Scope block
  NAS2D::Filesystem fs;  // Construct object (do not call init)
}  // Object goes out of scope and destructor is called
```

GDB backtrace (most recent call first):
```
#0  0x0000000000000000 in ?? ()
#1  0x00007ffff6a2815a in PHYSFS_setErrorCode ()
   from /usr/local/lib/libphysfs.so.1
#2  0x00007ffff6a2958f in PHYSFS_deinit () from /usr/local/lib/libphysfs.so.1
#3  0x0000000000408c09 in NAS2D::Filesystem::~Filesystem (this=0x7fffffffcfc0)
    at src/Filesystem.cpp:50
```

----

Suggested fix:
Move the contents of `init` into the constructor. That way the object is always fully formed, and can be safely destructed.

Side note: As the `Utility` class template was updated to allow passing arbitrary parameters to the constructor, this should be easier to manage now.

Related: Merging `init` into the constructor may also allow for elimination of the `FILESYSTEM_INITIALIZED` checks. If the object is fully constructed, and so there is a `this` object to refer to, then initialization has already run. Access to `Filesystem` methods will then be statically checked due to the program's structure. Therefor there will be no need for runtime checks.

**Comment by ldicker83** [comment]
I would much rather set a flag in the code. I'm not particularly keen on flaggy code like that but moving the init code into the c'tor seems like it would be bad form.

**Comment by ldicker83** [comment]
As a followup to my previous comments, PhysicsFS has exactly the sort of function we would need to avoid flaggy code:

https://icculus.org/physfs/docs/html/physfs_8h.html#a6154be7bfa5d924476356a9fbd0d8223

**Comment by DanRStevens** [comment]
I would say moving to single stage initialization is good form. Two stage initialization is a rather old practice, which has fallen out of favor. With two stage initialization you end up with these partially formed zombie objects, where it's not clear if it's safe to call any particular method on them. That's exactly what happened here.

The new style gaining popularity is to simply init the object fully in the constructor, and throw from the constructor if there was a problem. It's perfectly acceptable and safe to throw from a constructor. In many ways it's desirable, since you often want to prevent the formation of a partially constructed object, which throwing will do.

----

At any rate, merging the `init` method into the constructor would make `Filesystem` not default constructible. That means the change is currently blocked by the problem discovered in #114. There is a potential solution in #115, but it may need a bit more work.

----

Quite right about the `PHYSFS_isInit()` method. That would be a simple way to replace the `FILESYSTEM_INITIALIZED` flag.

----

Good idea to report the failure case to the PhysFS developer. I don't normally think to do that.
